crates/cranelift/src/compiler.rs

```diff
@@ -675,6 +726,61 @@ impl Compiler {
         builder.finalize();
     }
 
+    /// Loads the current fuel counter from `VMRuntimeLimits` into the fuel counter pinned reg.
+    fn reload_fuel(
+        &self,
+        offsets: &wasmtime_environ::VMOffsets<u8>,
+        builder: &mut FunctionBuilder,
+        vmctx: ir::Value,
+    ) {
+        let isa = &*self.isa;
+        let pointer_type = isa.pointer_type();
+        let vmctx_runtime_limits_offset = i32::try_from(offsets.vmctx_runtime_limits()).unwrap();
+        let vmruntime_limits_fuel_consumed_offset =
+            i32::try_from(offsets.vmruntime_limits_fuel_consumed()).unwrap();
+        let runtime_limits_ptr = builder.ins().load(
+            pointer_type,
+            ir::MemFlags::trusted(),
+            vmctx,
+            vmctx_runtime_limits_offset,
+        );
+        let new_fuel_consumed = builder.ins().load(
+            ir::types::I64,
+            ir::MemFlags::trusted(),
+            runtime_limits_ptr,
+            vmruntime_limits_fuel_consumed_offset,
+        );
+        builder.ins().set_pinned_reg(new_fuel_consumed);
+    }
+
+    /// Saves the current fuel counter saved in a pinned register into `VMRuntimeLimits` in memory.
+    fn spill_fuel(
+        &self,
+        offsets: &wasmtime_environ::VMOffsets<u8>,
+        builder: &mut FunctionBuilder,
+        vmctx: ir::Value,
+    ) {
+        let isa = &*self.isa;
+        let pointer_type = isa.pointer_type();
+        let fuel_consumed = builder.ins().get_pinned_reg(ir::types::I64);
+
+        let vmctx_runtime_limits_offset = i32::try_from(offsets.vmctx_runtime_limits()).unwrap();
+        let vmruntime_limits_fuel_consumed_offset =
+            i32::try_from(offsets.vmruntime_limits_fuel_consumed()).unwrap();
+        let runtime_limits_ptr = builder.ins().load(
+            pointer_type,
+            ir::MemFlags::trusted(),
+            vmctx,
+            vmctx_runtime_limits_offset,
+        );
+        builder.ins().store(
+            ir::MemFlags::trusted(),
+            fuel_consumed,
+            runtime_limits_ptr,
+            vmruntime_limits_fuel_consumed_offset,
+        );
+    }
```

Comment:
Should this also check fuel during spilling? Otherwise I think nondeterminism might creep in where host functions could execute when the wasm module has no fuel left?